vaadin-multi-select-list-mixin.html

```diff
@@ -0,0 +1,148 @@
+          observer: '_multipleChanged'
+        },
+
+        /**
+         * Array of indexes of the items selected in the items array
+         */
+        selectedValues: {
+          type: Array,
+          notify: true,
+          value: function() {
+            return [];
+          }
+        }
+      };
+    }
+
+    static get observers() {
+      return [
+        `_enhanceMultipleItems(items, multiple, selected, selectedValues, selectedValues.*)`
+      ];
+    }
+
+    ready() {
+      // Should be attached before click listener in list-mixin
+      this.addEventListener('click', e => this._onMultipleClick(e));
+
+      super.ready();
+      this.addEventListener('keydown', e => this._onMultipleKeydown(e));
+    }
+
+    _enhanceMultipleItems(items, multiple, selected, selectedValues) {
+      if (!items || !multiple) {
+        return;
+      }
+
+      if (selectedValues) {
+        const selectedItems = selectedValues.map(selectedId => items[selectedId]);
+        items.forEach(item => item.selected = selectedItems.indexOf(item) !== -1);
+      }
+
+      const lastSelectedItem = this.selectedValues.slice(-1);
+      if (lastSelectedItem && !lastSelectedItem.disabled) {
+        this._scrollToItem(lastSelectedItem);
+      }
+    }
+
+    _onMultipleClick(event) {
+      const item = this._filterItems(event.composedPath())[0];
+      const idx = item && !item.disabled ? this.items.indexOf(item) : -1;
+      if (idx < 0 || !this.multiple) {
+        return;
+      }
+
+      event.preventDefault();
+      if (event.shiftKey) {
+        const lastSelected = this.selectedValues.slice(-1);
+        if (lastSelected !== undefined && lastSelected === idx) {
+          return;
+        }
+        if (this.selectedValues.length) {
+          this.selectedValues = this._getAvailableIndexes(
+            Math.min(lastSelected, idx),
+            Math.max(lastSelected, idx),
+            item => !item.disabled
+          );
+        } else {
+          this.selectedValues = [idx];
+        }
+      } else if (event.metaKey || event.ctrlKey) {
+        this.__updateSelectedValues(idx);
+      } else {
+        this.selectedValues = [idx];
+      }
+    }
+
+    __updateSelectedValues(idx) {
+      if (this.selectedValues.indexOf(idx) !== -1) {
+        this.splice('selectedValues', this.selectedValues.indexOf(idx), 1);
+      } else {
+        this.push('selectedValues', idx);
+      }
+    }
+
+    _onMultipleKeydown(event) {
+      if (!this.multiple || !event.shiftKey) {
+        return;
+      }
+
+      const key = event.key.replace(/^Arrow/, '');
+
+      const focusedItem = this.items.filter(item => item.getAttribute('tabindex') === '0')[0];
+      const idx = this.items.indexOf(focusedItem);
+
+      if ((this._vertical && key === 'Up' || !this._vertical && key === 'Left') ||
+          (this._vertical && key === 'Down' || !this._vertical && key === 'Right')) {
+        this.__updateSelectedValues(idx);
```

Comment:
This should also work similarly to shift-clicking. It should select everything between the focused item and the item which was last active before the first shift-key press. Try going back-and-forth with arrow keys + shift.